### PR TITLE
security: close DNS-rebind SSRF residual in git remote dialing

### LINKED ISSUE #481: security: close DNS-rebind SSRF residual in git remote dialing (dial-time resolved-IP check)
Follow-up to #475 / PR #480.

## Context
`validateRemoteHost` (added in #480, `pkg/source/git/git.go`) rejects private/loopback/link-local/internal **IP literals and hostname suffixes** (`localhost`, `*.local`, `*.internal`) before dialing. This was the documented scope of #475.

It deliberately does **not** resolve DNS — so a public-looking hostname that resolves (or later re-resolves — a classic DNS-rebind/TOCTOU) to a private/loopback IP still passes the URL-host gate and gets dialed. CodeRabbit flagged this as an accepted residual on PR #480; we documented the tradeoff in the code comment and PR body.

## Proposed hardening (defense-in-depth)
Validate the **resolved connection IP at dial time**, not just the URL host:
- Install a custom transport for go-git's remote client (`go-git` `client.InstallProtocol` / a custom `http.Transport` with `net.Dialer.Control`) whose `Control` callback rejects any resolved address in the same private/loopback/link-local/unspecified/multicast set that `validateRemoteHost` covers.
- This closes the DNS-rebind gap for **all** remote dialing that shares the transport — branch listing **and** the poll/clone paths — so the guard becomes complete rather than literal-only.

## Constraints / gotchas
- `client.InstallProtocol` is **process-global** — must be installed once at startup and must not regress the existing poll/clone dialing (timeouts, proxy support via `HTTPS_PROXY`, existing auth). Keep the current literal-host `validateRemoteHost` as the cheap first gate; the dialer check is the second layer.
- Must respect the configured outbound proxy (do not bypass `HTTPS_PROXY`); when a proxy is in use the effective egress may differ — decide whether the check applies pre- or post-proxy.
- SSH/git protocols dial differently from http(s); scope which schemes get the dialer guard (http/https is the exploitable surface for the exfil path).
- Add tests: a hostname that resolves to a private IP must be rejected at dial time (inject resolution via a test dialer/hosts stub, no real network).

Not blocking #480 — that PR fully addresses the literal-host scope of #475.

### PR DESCRIPTION
**Pull request by dr14-make**
## Summary

Closes #481. Follow-up to #475 / PR #480.

`validateRemoteHost` (`pkg/source/git/git.go`) rejects loopback/private/link-local/internal **IP literals and hostname suffixes** before dialing a git remote — but only on the `ListBranches` path, and it does not resolve DNS. A public-looking hostname that resolves (or later re-resolves — a DNS-rebind/TOCTOU) to a private/loopback IP still passed that gate and got dialed, and the poll/clone/pull path had no host check at all. This PR adds the dial-time layer the issue asked for, covering both.

## What changed

- **`internal/gitops/dialguard.go`** (new) — `InstallSSRFGuardedTransport()` installs a process-wide go-git HTTP(S) transport (via `client.InstallProtocol("http"/"https", ...)`) whose `DialContext`:
  1. Resolves the hostname itself (or accepts an IP literal directly).
  2. Rejects the *whole* dial if **any** resolved candidate falls in `IsBlockedIP`'s loopback/private/link-local/unspecified/multicast set — a mix of public and private answers is itself treated as hostile/misconfigured DNS, so the "safe" candidates aren't used either.
  3. Connects directly to the validated IP (never by re-resolving the hostname a second time), so there is no window between "checked" and "connected" for a rebind to land in. Falls through to the next candidate on a dial error (proven by a dedicated test — see below).
  - Installed once via `sync.Once` + package `init()`, so every binary that imports `internal/gitops` (both `pkg/source/git` and `pkg/taskset`, confirmed the only two importers) gets the guard automatically — no call-site wiring to forget. This covers `ListBranches` **and** the poll/clone/pull paths, which `validateRemoteHost` alone never did.
  - Scoped to `http`/`https` only. **Not covered**: `ssh://` and `git://` schemes, which dial through go-git's own separate transports with no shared choke point — `git://` in particular has no injectable dialer at all in go-git, so guarding it would mean forking/reimplementing that transport. Filed as #486 since it's bigger than this issue's scope.
  - Preserves `http.DefaultTransport`'s 30s dial timeout / 30s keepalive explicitly (a code-review pass caught this being silently dropped in an earlier revision — see Review section) and falls back to a fresh `*http.Transport` rather than panicking if `http.DefaultTransport` is ever not a `*http.Transport` (defensive, since this runs from an unconditional `init()`).
  - When an outbound proxy is configured (`HTTPS_PROXY`), the dialer connects to the proxy's address, not the tunnelled target — documented in the function comment as the accepted tradeoff.
- **`pkg/source/git/git.go`** — `validateRemoteHost`'s IP-literal branch now calls the shared `gitops.IsBlockedIP` instead of duplicating the same six-condition check, so the two guard layers can't drift apart.
- **`docs/current-state.md`** — describes both guard layers, their actual scope (including that `validateRemoteHost` only ever runs on the `ListBranches` path), and the `ssh://`/`git://` gap.

## Test plan

Pure backend network-dial logic inside git remote handling with no user-facing trigger — not reachable via Playwright e2e (no UI action causes "a git remote's hostname resolves via DNS to an internal IP"; simulating that would need control over DNS resolution, which e2e can't provide). Regression coverage is unit-level:

- [x] `TestGuardedDialContext_RejectsResolvedPrivateIP` — the core ask: a hostname that isn't itself suspicious resolves to a private IP (169.254.169.254) → rejected before any dial is attempted.
- [x] `TestGuardedDialContext_RejectsWhenAnyCandidateBlocked` — mixed public+private resolution → rejected (fail-closed policy).
- [x] `TestGuardedDialContext_AllowsResolvedPublicIP` — happy path dials the *resolved* IP directly, proving no second (rebindable) hostname lookup happens.
- [x] `TestGuardedDialContext_FallsBackOnDialError` — first candidate's dial errors, second succeeds; proves the fallback loop is actually exercised, not just validated-and-never-tried.
- [x] `TestGuardedDialContext_IPLiteralSkipsResolver` / `_ResolveError` — edge cases.
- [x] `TestIsBlockedIP` — table test covering v4/v6 loopback, private (10/172.16/192.168), link-local, unique-local v6, unspecified, multicast, and IPv4-mapped-IPv6 variants of each.
- [x] `TestInstallSSRFGuardedTransport_Idempotent` / `_RejectsRealLoopbackViaGoGit` — repeated install doesn't panic; and a real go-git `ListContext` call against a real loopback `httptest.Server` is rejected, proving the guard is actually wired into go-git's request path (not just unit-tested in isolation).
- [x] `TestDialSettings_MatchHTTPDefaultTransport` — pins the restored 30s timeout/keepalive.
- [x] Existing `pkg/source/git` SSRF suite still green — confirms no regression to the literal-host guard or the public-host dial path.
- [x] `make build`, `go vet ./...`, `make format-check`, `make lint`, `go test ./internal/gitops/... ./pkg/source/... -race` all clean.
- [x] `make test` — full suite green except one pre-existing, unrelated environmental failure: `pkg/ipc TestControl_TaskTest_HappyPath` fails trying to download the Deno release binary (blocked by this sandbox's egress policy). Verified this fails identically on `origin/main` before this change.
- [ ] `make test-e2e` — skipped for the same environmental reason (the e2e harness boots the real daemon, which needs the Deno binary); also not applicable since this change has no UI-reachable surface.

## Review

Ran `/security-review`: found that `git://` scheme remotes bypass both guard layers entirely (filed as #486, doc/comments updated to state this instead of implying complete coverage).

Ran `/code-review` (8 finder angles): found and fixed a real regression — `dialTCP` was dialing with a zero-value `net.Dialer{}`, silently dropping the 30s timeout/keepalive `http.DefaultTransport`'s original dialer had, which would turn a dial to a black-holed/firewalled remote into an OS-level-timeout hang instead of a 30s failure. Also hardened the `http.DefaultTransport` type assertion against panicking, added a test proving the transport is really wired into go-git (not just unit-tested), and added a test for the previously-untested multi-candidate fallback path. Documented (not fixed) a narrow 6to4/Teredo tunnelling-address gap in `IsBlockedIP` — requires non-default OS tunnel routing to exploit.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_01Qw797yF42X1uSpMJSUUFzJ

## Summary by CodeRabbit

* **Bug Fixes**
  * Strengthened protections for Git remote fetching by blocking loopback, private, link-local, multicast, unspecified, and CGNAT IP ranges (including DNS rebind scenarios).
  * Tightened connection behavior to validate resolved addresses before dialing; preserves correct rejection while allowing only specifically configured proxy destinations.
* **Documentation**
  * Updated current-state documentation to clarify SSRF protection scope, coverage, and remaining gaps for non-HTTP(S) remotes.
* **Tests**
  * Expanded test coverage for IP-blocking logic, fail-closed dialing, proxy exemption behavior, transport installation, and an end-to-end go-git rejection check.

### CODE REVIEW

**Comment by dr14-make** [comment]
Ran `/security-review` on this diff. One confirmed finding, addressed:

**`git://` scheme remotes bypass both SSRF guard layers.** `pkg/taskset`'s `ValidateRefURL` accepts `git` as a valid `ref.url` scheme, but go-git's native git-protocol transport dials via a hardcoded `net.Dial` with no injectable hook — so a `git://` remote gets zero host validation, at either `validateRemoteHost` or this PR's new dial-time guard. Properly closing it means either rejecting the `git` scheme in taskset config validation or building a dedicated guarded transport for it (go-git's git-protocol client would need to be forked/reimplemented) — both bigger than this issue's (#481) explicitly-scoped surface (http/https). Filed as #486, and tightened the doc/code comments in this PR (8bde7b6) so they state the actual scope instead of implying complete coverage.

Everything else checked out: no TOCTOU window (resolves once, dials the validated IP directly), fail-closed on mixed public/private DNS answers, no silent unguarded fallback on resolver/dial errors, `sync.Once` install is race-free.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by dr14-make** [comment]
Marked ready for review. Summary of touched files:

- `internal/gitops/dialguard.go` (new) — process-wide dial-time SSRF guard for go-git's HTTP(S) transport.
- `internal/gitops/dialguard_test.go` (new) — unit tests, including one that drives a real go-git call against a real loopback server to prove the guard is wired into go-git's actual request path.
- `pkg/source/git/git.go` — dedup: `validateRemoteHost` now shares `gitops.IsBlockedIP` instead of a second copy of the same check.
- `docs/current-state.md` — documents the two guard layers and their actual scope (including the known `ssh://`/`git://` gap, tracked separately as #486).

Closes #481. Ready for human review/merge.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: internal/gitops/dialguard.go [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+		} else {
+			t = &stdhttp.Transport{Proxy: stdhttp.ProxyFromEnvironment}
+		}
+		t.DialContext = guardedDialContext
+
+		c := &stdhttp.Client{Transport: t}
+		client.InstallProtocol("http", githttp.NewClient(c))
+		client.InstallProtocol("https", githttp.NewClient(c))
+	})
+}
```
[the remark this answers is not in the document]
Comment (dr14-make) [reply internal/gitops/dialguard.go:224]: Confirmed and fixed in 545eda8 — `guardedDialContext` now exempts a dial target matching the configured HTTPS_PROXY/HTTP_PROXY host from the block check (dialed as-is, unresolved), since that hop is an operator-configured trusted egress path and very commonly on a private/loopback address itself. Added `TestGuardedDialContext_ExemptsConfiguredProxyHost` and `TestGuardedDialContext_NonProxyPrivateHostStillBlocked` to cover it.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by dr14-make** [comment]
Addressed both CodeRabbit findings in 545eda8:
- **Major** (proxy blocklist): fixed — see reply on the inline thread.
- **Nitpick** (CGNAT `100.64.0.0/10` not covered by `IsBlockedIP`): fixed, with test cases at the range boundaries.

Also caught and fixed a test-isolation issue the proxy fix introduced: this sandbox's own ambient `HTTPS_PROXY` happens to be `127.0.0.1`-based, which was making the "real go-git wiring" test's loopback server look like the configured proxy and silently pass for the wrong reason. All tests green, `-race` clean.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by dr14-make** [comment]
Addressed the `withStubs` parallel-safety nitpick in 0e9f24a.

---
_Generated by [Claude Code](https://claude.ai/code)_
